### DISCUSSION TITLE
Proposal for the naming of "Facade Service" and "App Service"

### DISCUSSION THREAD

**Discussion by rr-appadaptive** [FilledStacks/stacked#857] [Ideas]
Microservices and loose coupling are essential parts and paradigms of our software-architecture. In that context we have two major "players" that are comparable to your "Facade Service" and "App Service". The ones which does the work in our concept are the Workers (e. g. CustomerDataWorker, SqlServerWorker...). The ones which take care of the results of the workers and handle the orchestration of the services are the managers (e. g. MqttMessagingManager, CustomerManager, ModalDialogManager...). As you might ask, whether we do not use the word service anymore. We still do, but only for things like "recoursiveBomResolutionService", "ProductWeightCalculationService" etc.

Hope it helps.

**Comment by AmirC123** [comment]
In my use of the architecture I ended up using Data Access Objects (dao) for "Facade Service", but after thinking about it for a bit I think a better name would be "Access Point". 
So in your example it would be CustomerDataAccessPoint, SqalServerAccessPoint, etc. I think that makes more sense semantically for an app's service that abstracts some outside api/dependency.

It might be a good idea to keep "Service" for "App Service", so in your example it would still make sense to call it MqttMessagingService and ModalDialogService, though I admit CustomerService specifically can be a bit confusing because that's typically a whole other thing

**Comment by FilledStacks** [comment]
Hey @rr-appadaptive ,

Thanks for the proposal and the input. The naming is more for what "we" refer to, meaning the Stacked community. I want to stick with the word "service" in order to reduce the amount of bandwidth required when consuming or learning about Stacked. 

I thought I would spilt services into 2 to ensure that we cover all basis. This doesn't mean you can't have Workers or Managers. In fact, one of the ideas I played with was Manager as well. The reason I didn't go with it was the same as above. Personally, when I learn about some systems there are so many parts to it and I've rarely found it useful to have that level of abstraction. Granted I've not gone to the uber large scale where those Patterns come from.

Overall, as I educate and share Stacked, I always look for things that I can remove from the material and make the framework as easy to adapt as possible. With that being said. We have been working on providing a layer on-top of Stacked for the larger teams that are building big coded bases. 

It's called Clear Business Logic, I posted the idea of it a while ago (You can read about it [here](https://filledstacks.notion.site/Stacked-Clear-Business-Logic-CBL-c2d2970182094fbd960b98cd2ff230c1), it's not updated with the latest ideas). In this proposal there is a place for additional constructs other than services, Workers, Managers, that would make this even more scalable. 

One other thing to mention is that we're thinking more about what we call the service instead of "Facade Service" we have "X Service" that makes it easier to understand and remember. I used Facade service because of the popular [Facade Pattern](https://en.wikipedia.org/wiki/Facade_pattern) and I use App Service so that I can selfishly say in my tutorials and docs, "Write your App Logic in you App Service" 😄 

I think a lot about the developer experience, it's something I'm quite passionate about. So far simplifying, removing cogs, getting direct to core of the idea has proven to be useful. I really appreciate that you care enough to write this proposal, it means a lot to us. 

Makes putting in the hours into the project easier to pull off. 

Thanks @AmirC123 as well for giving your input. I agree with you. There's definitely a change we can make, the one I dislike the most is the "Facade Service". App Service I think is good, but I've thought about SingleResponsibility Service, which is way too long. Or "Wrapper" service, but then Wrapper kind of makes it feel like you can't do anything besides wrapping an existing package. Which is not the case. 

Let me know what you think.

**Reply by AmirC123** [reply]
Thanks for the detailed reply. I'm very much on board with your philosophy. The reason I chose Stacked is because it's the only framework that takes the effort to create readable code and reduce the mental overhead of figuring out "what's going on". With Stacked you can largely have that experience of simply "spelling out what you want" and it all makes sense. You don't *need* to know what a Provider is or all the ways you can use it, and you don't *need* a 10hrs course to figure out the BLoC concepts. You can understand any code without needing to know the framework. I think changing "notifyListeners" to "rebuildUi" was smart for this reason too.

One of the main pain point of any OOP coding is "what should go where". I feel like the those State vs Facade Service vs App Service is where some of this mental load is present in Stacked. 

Facade Service / Access Point is where I would abstract out outside APIs so I can mock them in necessary, or change them if I choose to. Conceptually this makes sense to me (though I agree with you that "facade service" is unpleasantly named, even though it aptly describes what it is. Hence my "access point" suggestion. I agree with you about disliking Wrapper. It also makes it sound like writing pointless code when it isn't).

As far as keeping the word "Service" for both: let say I have a "UserService" that manages the currently logged in user. It probably uses some Facade Service to do the actual CRUD in the database.  Wouldn't calling that facade service "UsersDatabaseService" be less readable than calling it "UsersDatabaseAccessPoint" as far as using semantics to reduce the mental load of figuring out the code, and help the developer more easily know "what goes where"?

Another thing for me is State vs App Service: How do I differentiate what should go where? "App Logic in App Service" is confusing because "App Logic" is too broad of a term. I would definitely need *some* logic in my State as well, so how much logic is too much logic for State? 

The way I'm trying to think about it, and hopefully you can tell me if I'm on the right track: would it be accurate to say that "State" should only have functions that ends with rebuildUi? Aka only functions that change the state should be in State, and since UI is a function of the state it should trigger a UI rebuild (or navigate to a different place in the app). 

And then everything else should go in the App Service? 

Is that a good way of looking at it?

**Comment by rr-appadaptive** [comment]
Hi Amir,

Your "stacked" paradigm is very convincing for me. Although I have been developing software for almost 40 years, apart from OOP, only Domain Driven Design (Eric Evans 2004) has really meant a quantum leap for what I do.

That's why I picked up Eric Evans' book again for a long time and was amazed how appropriate the diagram on the first page still is. (see the attached PDF).

I wish there were only services in Stacked (no further distinction). This leaves me the necessary degree of freedom and does not force me to decide what kind of service it should be. In addition, I would find it very helpful if there were repositories (very close to the idea of access points), factories and maybe even entities analogous to DDD.

Best Regards,

Regine Rosewich
—
fon: +49-7645-9130-40
fax: +49-7645-9130-42
mo: +49-171-7791167
eM: ***@***.******@***.***>

[signature_836126763]

AppAdaptive GmbH
Ziegelstraße 13
88212 Ravensburg
Tel.: +49 (751) 36605-0
Fax: +49 (751) 36605-20
________________________________________________________________
Handelsregister HRB 731 585 • Amtsgericht Ulm • Sitz der Gesellschaft: Ravensburg •
Geschäftsführung: Christian Steidle, Regine Rosewich

This e-mail may contain confidential and/or privileged information.
If you are not the intended recipient (or have received this e-mail
in error) please notify the sender immediately and destroy this e-mail.
Any unauthorized copying, disclosure or distribution on the material
in this e-mail is strictly forbidden. We exclude liability for any damage
caused by computer-viruses or the like insofar as we did not act willful
or grossly negligent. We advise you to carry out your own Virus-Check
before you open any of the attachments.





Von: Amir Cohen ***@***.***>
Datum: Mittwoch, 8. Februar 2023 um 15:34
An: FilledStacks/stacked ***@***.***>
Cc: Regine Rosewich ***@***.***>, Mention ***@***.***>
Betreff: Re: [FilledStacks/stacked] Proposal for the naming of "Facade Service" and "App Service" (Discussion #857)

Thanks for the detailed reply. I'm very much on board with your philosophy. The reason I chose Stacked is because it's the only framework that takes the effort to create readable code and reduce the mental overhead of figuring out "what's going on". With Stacked you can largely have that experience of simply "spelling out what you want" and it all makes sense. You don't need to know what a Provider is or all the ways you can use it, and you don't need a 10hrs course to figure out the BLoC concepts. You can understand any code without needing to know the framework. I think changing "notifyListeners" to "rebuildUi" was smart for this reason too.

One of the main pain point of any OOP coding is "what should go where". I feel like the those State vs Facade Service vs App Service is where some of this mental load is present in Stacked.

Facade Service / Access Point is where I would abstract out outside APIs so I can mock them in necessary, or change them if I choose to. Conceptually this makes sense to me (though I agree with you that "facade service" is unpleasantly named, even though it aptly describes what it is. Hence my "access point" suggestion. I agree with you about disliking Wrapper. It also makes it sound like writing pointless code when it isn't).

As far as keeping the word "Service" for both: let say I have a "UserService" that manages the currently logged in user. It probably uses some Facade Service to do the actual CRUD in the database. Wouldn't calling that facade service "UsersDatabaseService" be less readable than calling it "UsersDatabaseAccessPoint" as far as using semantics to reduce the mental load of figuring out the code, and help the developer more easily know "what goes where"?

Another thing for me is State vs App Service: How do I differentiate what should go where? "App Logic in App Service" is confusing because "App Logic" is too broad of a term. I would definitely need some logic in my State as well, so how much logic is too much logic for State?

The way I'm trying to think about it, and hopefully you can tell me if I'm on the right track: would it be accurate to say that "State" should only have functions that ends with rebuildUi? Aka only functions that change the state should be in State, and since UI is a function of the state it should trigger a UI rebuild (or navigate to a different place in the app).

And then everything else should go in the App Service?

Is that a good way of looking at it?

**Comment by ferrarafer** [comment]
> I wish there were only services in Stacked (no further distinction). This leaves me the necessary degree of freedom and does not force me to decide what kind of service it should be.

Hi @rr-appadaptive ! In fact, you don't have to choose because the distinction is only conceptual. Both are a simple Dart class, @FilledStacks  divided them into two groups or types to explain the aim for which they are going to be created. We can see it as a distinction for educational purpose.
